include/revng/ADT/GenericGraph.h

```diff
@@ -996,6 +996,124 @@ class MutableEdgeNode
+    }
+    static UndirectedChildIterator<SuccessorIterator, PredecessorIterator>
+    end(llvm::iterator_range<SuccessorIterator> Successors,
+        llvm::iterator_range<PredecessorIterator> Predecessors) {
+      return UndirectedChildIterator<
+        SuccessorIterator,
+        PredecessorIterator>(std::end(Successors),
+                             std::end(Successors),
+                             std::end(Predecessors),
+                             std::end(Predecessors));
+    }
+
+    using UndirectedChildIterator::iterator_facade_base::operator++;
+
+    UndirectedChildIterator &operator++() {
+      if (SBegin != SEnd) {
+        ++SBegin;
+        return *this;
+      }
+
+      if (PBegin != PEnd) {
+        ++PBegin;
+        return *this;
+      }
+
+      revng_abort("Attempted to increment an end iterator!");
+    }
+
+    decltype(auto) operator*() {
+      if (SBegin != SEnd)
+        return *SBegin;
+
+      if (PBegin != PEnd)
+        return *PBegin;
+
+      revng_abort("Attempted to dereference an end iterator!");
+    }
+    decltype(auto) operator*() const {
+      if (SBegin != SEnd)
+        return *SBegin;
+
+      if (PBegin != PEnd)
+        return *PBegin;
+
+      revng_abort("Attempted to dereference an end iterator!");
+    }
+
+    // Proxy object which extends the lifetime of the `EdgeView` object returned
+    // by the underlying `map_iterator`
+    struct Proxy {
+      Proxy(EdgeView Value) : Temporary(std::move(Value)) {}
+      EdgeView *const operator->() { return &Temporary; }
+      EdgeView const *const operator->() const { return &Temporary; }
+
+    private:
+      EdgeView Temporary;
+    };
+    Proxy operator->() { return operator*(); }
+    Proxy operator->() const { return operator*(); }
+
+    bool operator==(const UndirectedChildIterator &Another) const {
+      auto This = std::tie(SBegin, SEnd, PBegin, PEnd);
+      auto That = std::tie(Another.SBegin,
+                           Another.SEnd,
+                           Another.PBegin,
+                           Another.PEnd);
+      return This == That;
+    }
+  };
+
+public:
+  auto undirected_children() {
+    using ResultIterator = UndirectedChildIterator<SuccessorIterator,
+                                                   PredecessorIterator>;
+    return llvm::make_range(ResultIterator::begin(successors(), predecessors()),
+                            ResultIterator::end(successors(), predecessors()));
+  }
+  auto undirected_child_edges() {
+    using ResultIterator = UndirectedChildIterator<SuccessorEdgeIterator,
+                                                   PredecessorEdgeIterator>;
+    return llvm::make_range(ResultIterator::begin(successor_edges(),
+                                                  predecessor_edges()),
+                            ResultIterator::end(successor_edges(),
+                                                predecessor_edges()));
+  }
```

Comment:
I guess these methods could be in principle implemented just using `GraphTraits` and become available for any implementor of `GraphTraits` and not just `GenericGraph`, correct?

I assume this is reflected by the comment below:
```
// TODO: We should move the `Undirected` definition to a separate header, and
//       offer a default implementation for `GraphTraits<Undirected<T>>` for
//       any `T` for which you have a definition of both `GraphTraits<T>` and
//       `GraphTraits<Inverse<T>>`
```
If this is the case, this is fine.

Reply:
Yes, your observation is correct, even thought the comment below is indeed a bit outdated and refers to the previous implementation, so I'm updating this

Reply:
I added a specific, and more clear comment highlighting your observation with a TODO https://github.com/revng/revng/blob/f14ffdc32258b98198148f45162302e6cbecc7c1/include/revng/ADT/GenericGraph.h#L1001